File: backend/app/models/contacts.py
"""
Contacts module Pydantic models — customers, general contractors,
entity contacts, and job linking junctions.

Covers all request/response shapes for the Contacts management endpoints.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


# ── Shared Literals ───────────────────────────────────────────────

CUSTOMER_TYPES = Literal["residential", "commercial", "government", "other"]

GC_TRADE_TYPES = Literal[
    "general", "electrical", "plumbing", "hvac",
    "mechanical", "fire_protection", "low_voltage", "other",
]

ENTITY_TYPES = Literal["customer", "general_contractor", "supplier"]

CONTACT_ROLES = Literal[
    "owner", "property_manager", "tenant",
    "site_contact", "billing", "other",
]

GC_RELATIONSHIPS = Literal["they_are_gc", "we_hired_them"]


# ── Customer ──────────────────────────────────────────────────────

class CustomerCreate(BaseModel):
    """Create a new customer."""
    company_name: str | None = None
    first_name: str = Field(..., min_length=1, max_length=100)
    last_name: str = Field(..., min_length=1, max_length=100)
    phone: str | None = None
    email: str | None = None
    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    state: str | None = None
    zip: str | None = None
    customer_type: CUSTOMER_TYPES = "residential"
    notes: str | None = None


class CustomerUpdate(BaseModel):
    """Partial update for an existing customer."""
    company_name: str | None = None
    first_name: str | None = Field(default=None, min_length=1, max_length=100)
    last_name: str | None = Field(default=None, min_length=1, max_length=100)
    phone: str | None = None
    email: str | None = None
    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    state: str | None = None
    zip: str | None = None
    customer_type: CUSTOMER_TYPES | None = None
    notes: str | None = None


class CustomerResponse(BaseModel):
    """Full customer detail returned from the API."""
    id: int
    company_name: str | None = None
    first_name: str
    last_name: str
    display_name: str | None = None
    phone: str | None = None
    email: str | None = None
    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    state: str | None = None
    zip: str | None = None
    customer_type: str = "residential"
    notes: str | None = None
    is_active: bool = True
    created_at: datetime | None = None
    updated_at: datetime | None = None

    # Hydrated in detail endpoint
    contacts: list[EntityContactResponse] = Field(default_factory=list)
    job_count: int = 0


class CustomerListItem(BaseModel):
    """Lightweight customer for list views."""
    id: int
    company_name: str | None = None
    first_name: str
    last_name: str
    display_name: str | None = None
    phone: str | None = None
    email: str | None = None
    customer_type: str = "residential"
    is_active: bool = True
    job_count: int = 0
    contact_count: int = 0


# ── General Contractor ────────────────────────────────────────────

class GCCreate(BaseModel):
    """Create a new general contractor."""
    company_name: str = Field(..., min_length=1, max_length=200)
    gc_code: str = Field(..., min_length=1, max_length=20)
    license_number: str | None = None
    trade_type: GC_TRADE_TYPES = "general"
    phone: str | None = None
    email: str | None = None
    website: str | None = None
    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    state: str | None = None
    zip: str | None = None
    insurance_info: str | None = None
    notes: str | None = None


class GCUpdate(BaseModel):
    """Partial update for an existing GC."""
    company_name: str | None = Field(default=None, min_length=1, max_length=200)
    gc_code: str | None = Field(default=None, min_length=1, max_length=20)
    license_number: str | None = None
    trade_type: GC_TRADE_TYPES | None = None
    phone: str | None = None
    email: str | None = None
    website: str | None = None
    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    state: str | None = None
    zip: str | None = None
    insurance_info: str | None = None
    notes: str | None = None


class GCResponse(BaseModel):
    """Full GC detail returned from the API."""
    id: int
    company_name: str
    gc_code: str
    license_number: str | None = None
    trade_type: str = "general"
    phone: str | None = None
    email: str | None = None
    website: str | None = None
    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    state: str | None = None
    zip: str | None = None
    insurance_info: str | None = None
    notes: str | None = None
    is_active: bool = True
    created_at: datetime | None = None
    updated_at: datetime | None = None

    # Hydrated in detail endpoint
    contacts: list[EntityContactResponse] = Field(default_factory=list)
    job_count: int = 0


class GCListItem(BaseModel):
    """Lightweight GC for list views."""
    id: int
    company_name: str
    gc_code: str
    trade_type: str = "general"
    phone: str | None = None
    email: str | None = None
    is_active: bool = True
    job_count: int = 0
    contact_count: int = 0


# ── Entity Contact ────────────────────────────────────────────────

class EntityContactCreate(BaseModel):
    """Add a contact to a customer, GC, or supplier."""
    first_name: str = Field(..., min_length=1, max_length=100)
    last_name: str = Field(default="", max_length=100)
    role: str = Field(..., min_length=1, max_length=100)  # freetext
    phone: str = Field(..., min_length=1)  # mandatory
    email: str | None = None
    is_primary: bool = False
    notes: str | None = None


class EntityContactUpdate(BaseModel):
    """Partial update for an entity contact."""
    first_name: str | None = Field(default=None, min_length=1, max_length=100)
    last_name: str | None = Field(default=None, max_length=100)
    role: str | None = Field(default=None, min_length=1, max_length=100)
    phone: str | None = Field(default=None, min_length=1)
    email: str | None = None
    is_primary: bool | None = None
    notes: str | None = None


class EntityContactResponse(BaseModel):
    """Entity contact returned from the API."""
    id: int
    entity_type: str
    entity_id: int
    first_name: str
    last_name: str = ""
    role: str
    phone: str
    email: str | None = None
    is_primary: bool = False
    notes: str | None = None
    is_active: bool = True
    created_at: datetime | None = None
    updated_at: datetime | None = None


# ── Directory (Unified Contact Search) ────────────────────────────

class DirectoryContactResult(BaseModel):
    """Unified search result across all entity_contacts."""
    id: int
    first_name: str
    last_name: str = ""
    role: str
    phone: str
    email: str | None = None
    entity_type: str          # 'customer', 'general_contractor', 'supplier'
    entity_id: int
    entity_name: str = ""     # resolved parent name (customer display_name, gc company_name, supplier name)


# ── Job ↔ Customer Linking ────────────────────────────────────────

class JobCustomerCreate(BaseModel):
    """Link a customer to a job."""
    customer_id: int
    contact_role: CONTACT_ROLES = "owner"
    is_primary: bool = False
    notes: str | None = None


class JobCustomerResponse(BaseModel):
    """Job-customer link returned from the API."""
    id: int
    job_id: int
    customer_id: int
    contact_role: str = "owner"
    is_primary: bool = False
    notes: str | None = None
    created_at: datetime | None = None

    # Hydrated from customer
    customer_name: str | None = None
    company_name: str | None = None
    phone: str | None = None
    email: str | None = None


# ── Job ↔ GC Linking ─────────────────────────────────────────────

class JobGCCreate(BaseModel):
    """Link a general contractor to a job."""
    gc_id: int
    relationship: GC_RELATIONSHIPS
    contract_amount: float | None = None
    contract_number: str | None = None
    is_primary: bool = False
    notes: str | None = None


class JobGCResponse(BaseModel):
    """Job-GC link returned from the API."""
    id: int
    job_id: int
    gc_id: int
    relationship: str
    contract_amount: float | None = None
    contract_number: str | None = None
    is_primary: bool = False
    notes: str | None = None
    created_at: datetime | None = None

    # Hydrated from GC
    company_name: str | None = None
    gc_code: str | None = None
    trade_type: str | None = None
    phone: str | None = None
    email: str | None = None

File: src/lib/types/contacts.ts
/**
 * Contacts types — customers, general contractors, entity contacts (polymorphic),
 * job ↔ customer/GC linking.
 */

// ══════════════════════════════════════════════════════════════════
// CUSTOMERS
// ══════════════════════════════════════════════════════════════════

export type CustomerType = 'residential' | 'commercial' | 'government' | 'other';

export interface CustomerListItem {
  id: number;
  company_name: string | null;
  first_name: string;
  last_name: string;
  display_name: string;
  phone: string | null;
  email: string | null;
  customer_type: CustomerType;
  is_active: boolean;
  job_count: number;
  contact_count: number;
}

export type PaymentTerms = 'due_on_receipt' | 'net_15' | 'net_30' | 'net_45' | 'net_60' | 'custom';

export interface CustomerDetail {
  id: number;
  company_name: string | null;
  company_code: string | null;
  first_name: string;
  last_name: string;
  display_name: string;
  phone: string | null;
  email: string | null;
  address_line1: string | null;
  address_line2: string | null;
  city: string | null;
  state: string | null;
  zip: string | null;
  customer_type: CustomerType;
  notes: string | null;
  is_active: boolean;
  contacts: EntityContactResponse[];
  job_count: number;
  created_at: string | null;
  updated_at: string | null;
  // Billing
  billing_address_line1: string | null;
  billing_address_line2: string | null;
  billing_city: string | null;
  billing_state: string | null;
  billing_zip: string | null;
  payment_terms: PaymentTerms;
  tax_id: string | null;
  billing_email: string | null;
}

export interface CustomerCreate {
  company_name?: string | null;
  company_code?: string | null;
  first_name: string;
  last_name: string;
  phone?: string | null;
  email?: string | null;
  address_line1?: string | null;
  address_line2?: string | null;
  city?: string | null;
  state?: string | null;
  zip?: string | null;
  customer_type?: CustomerType;
  notes?: string | null;
  billing_address_line1?: string | null;
  billing_address_line2?: string | null;
  billing_city?: string | null;
  billing_state?: string | null;
  billing_zip?: string | null;
  payment_terms?: PaymentTerms;
  tax_id?: string | null;
  billing_email?: string | null;
}

export interface CustomerUpdate {
  company_name?: string | null;
  company_code?: string | null;
  first_name?: string;
  last_name?: string;
  phone?: string | null;
  email?: string | null;
  address_line1?: string | null;
  address_line2?: string | null;
  city?: string | null;
  state?: string | null;
  zip?: string | null;
  customer_type?: CustomerType;
  notes?: string | null;
  billing_address_line1?: string | null;
  billing_address_line2?: string | null;
  billing_city?: string | null;
  billing_state?: string | null;
  billing_zip?: string | null;
  payment_terms?: PaymentTerms;
  tax_id?: string | null;
  billing_email?: string | null;
}


// ══════════════════════════════════════════════════════════════════
// GENERAL CONTRACTORS
// ══════════════════════════════════════════════════════════════════

export type GCTradeType =
  | 'general'
  | 'electrical'
  | 'plumbing'
  | 'hvac'
  | 'mechanical'
  | 'fire_protection'
  | 'low_voltage'
  | 'other';

export interface GCListItem {
  id: number;
  company_name: string;
  gc_code: string;
  trade_type: GCTradeType;
  phone: string | null;
  email: string | null;
  is_active: boolean;
  job_count: number;
  contact_count: number;
}

export interface GCDetail {
  id: number;
  company_name: string;
  gc_code: string;
  license_number: string | null;
  trade_type: GCTradeType;
  phone: string | null;
  email: string | null;
  website: string | null;
  address_line1: string | null;
  address_line2: string | null;
  city: string | null;
  state: string | null;
  zip: string | null;
  insurance_info: string | null;
  notes: string | null;
  is_active: boolean;
  contacts: EntityContactResponse[];
  job_count: number;
  created_at: string | null;
  updated_at: string | null;
  // COI tracking
  coi_carrier: string | null;
  coi_policy_number: string | null;
  coi_expiry_date: string | null;
  coi_coverage_amount: number | null;
  coi_on_file: boolean;
  workers_comp_expiry: string | null;
  bonded: boolean;
  bond_amount: number | null;
}

export interface GCCreate {
  company_name: string;
  gc_code: string;
  license_number?: string | null;
  trade_type?: GCTradeType;
  phone?: string | null;
  email?: string | null;
  website?: string | null;
  address_line1?: string | null;
  address_line2?: string | null;
  city?: string | null;
  state?: string | null;
  zip?: string | null;
  insurance_info?: string | null;
  notes?: string | null;
  coi_carrier?: string | null;
  coi_policy_number?: string | null;
  coi_expiry_date?: string | null;
  coi_coverage_amount?: number | null;
  coi_on_file?: boolean;
  workers_comp_expiry?: string | null;
  bonded?: boolean;
  bond_amount?: number | null;
}

export interface GCUpdate {
  company_name?: string;
  gc_code?: string;
  license_number?: string | null;
  trade_type?: GCTradeType;
  phone?: string | null;
  email?: string | null;
  website?: string | null;
  address_line1?: string | null;
  address_line2?: string | null;
  city?: string | null;
  state?: string | null;
  zip?: string | null;
  insurance_info?: string | null;
  notes?: string | null;
  coi_carrier?: string | null;
  coi_policy_number?: string | null;
  coi_expiry_date?: string | null;
  coi_coverage_amount?: number | null;
  coi_on_file?: boolean;
  workers_comp_expiry?: string | null;
  bonded?: boolean;
  bond_amount?: number | null;
}


// ══════════════════════════════════════════════════════════════════
// ENTITY CONTACTS (polymorphic — customers, GCs, suppliers)
// ══════════════════════════════════════════════════════════════════

export type EntityType = 'customer' | 'general_contractor' | 'supplier';

export interface EntityContactResponse {
  id: number;
  entity_type: EntityType;
  entity_id: number;
  first_name: string;
  last_name: string;
  role: string;
  phone: string;
  email: string | null;
  is_primary: boolean;
  notes: string | null;
  is_active: boolean;
  created_at: string | null;
  updated_at: string | null;
}

export interface EntityContactCreate {
  first_name: string;
  last_name: string;
  role: string;
  phone: string;
  email?: string | null;
  is_primary?: boolean;
  notes?: string | null;
}

export interface EntityContactUpdate {
  first_name?: string;
  last_name?: string;
  role?: string;
  phone?: string;
  email?: string | null;
  is_primary?: boolean;
  notes?: string | null;
}

export interface DirectoryContactResult {
  id: number;
  first_name: string;
  last_name: string;
  role: string;
  phone: string;
  email: string | null;
  entity_type: EntityType;
  entity_id: number;
  entity_name: string;
}


// ══════════════════════════════════════════════════════════════════
// JOB ↔ CUSTOMER / GC LINKING
// ══════════════════════════════════════════════════════════════════

export type CustomerContactRole =
  | 'owner'
  | 'property_manager'
  | 'tenant'
  | 'site_contact'
  | 'billing'
  | 'other';

export type GCRelationship = 'they_are_gc' | 'we_hired_them';

export interface JobCustomerResponse {
  id: number;
  job_id: number;
  customer_id: number;
  customer_name: string;
  company_name: string | null;
  phone: string | null;
  email: string | null;
  contact_role: CustomerContactRole;
  is_primary: boolean;
  notes: string | null;
  created_at: string | null;
}

export interface JobCustomerCreate {
  customer_id: number;
  contact_role?: CustomerContactRole;
  is_primary?: boolean;
  notes?: string | null;
}

export interface JobGCResponse {
  id: number;
  job_id: number;
  gc_id: number;
  company_name: string;
  gc_code: string;
  trade_type: GCTradeType;
  phone: string | null;
  email: string | null;
  relationship: GCRelationship;
  contract_amount: number | null;
  contract_number: string | null;
  is_primary: boolean;
  notes: string | null;
  created_at: string | null;
}

export interface JobGCCreate {
  gc_id: number;
  relationship: GCRelationship;
  contract_amount?: number | null;
  contract_number?: string | null;
  is_primary?: boolean;
  notes?: string | null;
}

/** Job linked to a customer — used on the Customer Detail "Jobs" tab (reverse of JobCustomerResponse) */
export interface CustomerJobLink {
  id: number;
  job_id: number;
  customer_id: number;
  job_name: string;
  job_status: string;
  contact_role: CustomerContactRole;
  is_primary: boolean;
  notes: string | null;
  created_at: string | null;
}

/** Job linked to a GC — used on the Contractor Detail "Jobs" tab (reverse of JobGCResponse) */
export interface GCJobLink {
  id: number;
  job_id: number;
  gc_id: number;
  job_name: string;
  job_status: string;
  relationship: GCRelationship;
  contract_amount: number | null;
  contract_number: string | null;
  is_primary: boolean;
  notes: string | null;
  created_at: string | null;
}
